core/coreapi/coreapi.go

```diff
@@ -217,6 +216,10 @@ func (api *CoreAPI) WithOptions(opts ...options.ApiOption) (coreiface.CoreAPI, e
 
 	}
 
+	if !settings.FetchBlocks {
+		subApi.dag = dag.NewDAGService(bserv.New(subApi.blockstore, offlinexch.Exchange(subApi.blockstore)))
```

Comment:
`settings.FetchBlocks` needs to set `exchange`/`blocks` too, otherwise it still will be possible to use gateway to fetch stuff from the network with things like `/api/v0/block/get`